### ISSUE TITLE
[RFC] Improved imports

### ISSUE THREAD

**Issue by vegansk** [gcanti/fp-ts#143]
### RFC 1. Export defaults

What do you think about exporting the content of the modules as the default? For now, we should write imports like this:
```ts
import { Option, some, none } from 'fp-ts/lib/Option'
import * as option from 'fp-ts/lib/Option'
```

And with this changes, we can write it like this:
```ts
import option, { Option, some, none } from 'fp-ts/lib/Option'
```

### RFC 2. Module names

What's the need of the `lib` part in the module name? Isn't it better to have it like this:
```ts
import option, { Option, some, none } from 'fp-ts/Option'
```

**Comment by vegansk** [comment]
One more comment. If you write import like this:
```ts
import option, { Option, some, none } from 'fp-ts/lib/Option'
```
now, the compiler doesn't fail. It fails in runtime. Checked with typescript 2.3.4

**Comment by raveclassic** [comment]
My two cents here:
WS and other editors have impressive autoimport support but it just don't work with `* as option` syntax because there's no actual named export. So each time one has to manually write those imports, it's a bit inconvenient. 
Is there any possibility to have a named exports for a HKT?

**Comment by gcanti** [comment]
@vegansk @raveclassic sorry for the delay (busy with the 0.4.0 rewrite). Interesting suggestions, could you please elaborate though? In practice what should I do? Is there a repo setup somewhere that I can replicate?

**Comment by raveclassic** [comment]
@gcanti Well every setup will work for autoimports where there's an explicit named export like:
```typescript
//previous
export function some() {}
export function none() {}
//...
import * as option from 'fp-ts/lib/Option'; //always should be writte by hand

//would be great
export function some() {}
export function none() {}
//this is important for autoimport to work
export const option: HKT /* or smth to check interface */ = {
  some,
  none
};
//original request for default imports
export default option;

//...

import { option } from 'fp-ts/lib/Option'; //autoimported when using 'option' somewhere
import option from 'fp-ts/lib/Option'; //default import, still written by hand
```

**Comment by gcanti** [comment]
@raveclassic got it thanks, so could be just what I'm calling `proof` to check the implementation (example here https://github.com/gcanti/fp-ts/blob/650857691b94435f340d9b3662856ffef031dcab/src/Option.ts#L304) but currently is not exported right?

**Comment by davetayls** [comment]
My 2 cents on this if I may. I've found the same issues with importing so am with @raveclassic . I will end up doing something like this a lot

```typescript
import * as Option from 'fp-ts/lib/Option`;
type IOption = Option.Option;
```

I favour the uppercase `Option` here because that's often the style you see a lot of functional docs/libraries written and in the same way you will often see interfaces in TypeScript modules using the `I` at the beginning.

Hope this is constructive 😄

**Comment by raveclassic** [comment]
@gcanti If such proof can be passed to functions which accept a HKT (like `sequence`) then yes, it would be awesome

**Comment by gcanti** [comment]
@raveclassic yep! It's not always possible though. `Option` for example: if you want to use `sequence` it would be pretty handy 

```ts
import { sequence } from 'fp-ts/lib/Traversable'
import { option, some } from 'fp-ts/lib/Option' // proof of Option
import { array } from 'fp-ts/lib/Array' // proof of Array

const foo = sequence(option, array)([some(1)])
```

but if you want a monoid instance you have to call the `getMonoid` function  

```ts
import { getMonoid } from 'fp-ts/lib/Option'
import { monoidString } from 'fp-ts/lib/Monoid'

export const mymonoid = getMonoid(monoidString)
```

@davetayls isn't the same as

```ts
import { Option } from 'fp-ts/lib/Option'
// or
import { Option as IOption } from 'fp-ts/lib/Option'
```

**Comment by gcanti** [comment]
Maybe a better example is `Const`: I can certainly export the proof but it only contains the `Functor` and `Contravariant` instances. In order to get an `Applicative` instance you have to call `getApplicative`

**Comment by gcanti** [comment]
Anyway I find myself using `import * as option` only when I want to use `option` as a proof for functions like `sequence` so if I export all the proofs it seems a net gain, isn't it?

**Comment by raveclassic** [comment]
@gcanti Yes, it would be great!
But.. what name to pick for Const proof? `const` is reserved :)

**Comment by davetayls** [comment]
@gcanti ah sorry i should have included more in the example. This is over-simplified but hopefully shows the separate `Option` and `IOption`.

```typescript
import * as Option from 'fp-ts/lib/Option`;
type IOption = Option.Option;

const v: IOption<string> = Option.fromNullable(something);
```

Sounds like exporting the proofs as default would be a gain. It would make the example a lot nicer, especially if the `Option` interface isn't needed. Something like this I presume.

```typescript
import Option, { Option as IOption } from 'fp-ts/lib/Option`;
const v: IOption<string> = Option.fromNullable(something);
```

**Comment by raveclassic** [comment]
@davetayls proofs actually don't include helpers like `fromNullable` so that you have to either import it separately or with `import * as Option` syntax.

**Comment by davetayls** [comment]
oops, didn't see that thanks for the heads up 😄

**Comment by gcanti** [comment]
@raveclassic mhh.. `const_`? [Here](https://github.com/gcanti/fp-ts/blob/650857691b94435f340d9b3662856ffef031dcab/src/index.ts) you can find the current names exported (as whole modules) in `index.ts`

**Comment by raveclassic** [comment]
@gcanti Sounds good, at least IDE/editor will suggest it.
Unfortunately JS parser can't process `const'` which imo fits better and is more common in fp languages.

**Comment by gcanti** [comment]
closed this in 03ee0fa33b248383ae566d0a61d52049f484108b
